src/python/pants/option/parser.py

```diff
@@ -551,6 +557,11 @@ def record_option(value, rank, option_details=None):
     def check(val):
       if val is not None:
         choices = kwargs.get('choices')
+        # Reach into `enum` types to get the allowed `choices` if not explicitly set.
+        if choices is None and 'type' in kwargs:
+          type_arg = kwargs.get('type')
+          if hasattr(type_arg, 'iterate_enum_variants'):
```

Comment:
Can you check for `issubclass(type_arg, ChoiceDatatype)` instead? That's what you really intend, no?

Reply:
The `ChoiceDatatype` doesn't leave the body of the `enum` method -- if we made all enums mix in something else to mark them as an enum, we could do that, but currently an `enum` is just a datatype with some extra methods. I was thinking that checking for a method like this would be more "pythonic".

One way to do either of these is to define an empty `ChoicesMixin` class which `enum` classes mix in, which has a `get_all_variants()` abstractmethod, then check `isinstance(type_arg, ChoicesMixin)` -- this also means other things outside of `objects.py` that want to be option types with implicit `choices` can mix in `ChoicesMixin`. I can push a commit that shows what I'm talking about.

Reply:
I actually decided to keep using `hasattr` because functions are also used as `type` arguments, so `issubclass(type_arg, ChoicesMixin)` would actually just fail. This also allows functions to have implicit `choices` arguments by adding the `all_variants` attribute -- I've left a comment with this briefly described.

Reply:
Gotcha, this is fine as-is then.